verifier/src/x509/chain.rs

```diff
@@ -29,33 +30,89 @@ impl CertificateChain {
 
     /// Returning the signing key from the leaf certificate
     ///
-    /// The chain will be verified against the `trust_root` and the `unix_time`.
+    /// The chain will be verified against the `trust_anchor`, `unix_time`, and
+    /// `crls`.
+    ///
+    /// # Arguments
+    /// * `trust_anchor` - The trust anchor of the chain. This is often a
+    ///   previously known certificate.
+    /// * `unix_time` - The duration since
+    ///   [`UNIX_EPOCH`](https://doc.rust-lang.org/std/time/constant.UNIX_EPOCH.html).
+    ///   This is expected to be generated by the caller using:
+    ///     ```ignore
+    ///     SystemTime::now().duration_since(UNIX_EPOCH)
+    ///     ```
+    ///   or equivalent
+    /// * `crls` - A list of certificate revocation lists that will be used to
+    ///   reject revoked certificates.
     pub fn signing_key(
         &self,
-        trust_root: VerifiedCertificate,
+        trust_anchor: &VerifiedCertificate,
         unix_time: Duration,
+        crls: &[UnverifiedCrl],
     ) -> Result<VerifyingKey> {
-        let mut key = trust_root.public_key();
+        let mut signing_cert = trust_anchor.clone();
         for cert in &self.certificates {
+            let key = signing_cert.public_key();
             let verified_cert = cert.verify(&key, unix_time)?;
-            key = verified_cert.public_key();
+            verify_certificate_not_revoked(&verified_cert, &signing_cert, crls, unix_time)?;
+            signing_cert = verified_cert;
+        }
+        Ok(signing_cert.public_key())
+    }
+}
+
+/// Convert a series PEM-encoded certificates into a [`CertificateChain`].
+///
+/// The PEMs must be in order from the trust anchor to the leaf certificate.
+impl TryFrom<&[&str]> for CertificateChain {
+    type Error = Error;
+
+    fn try_from(pems: &[&str]) -> ::core::result::Result<Self, Self::Error> {
+        let certificates = pems
+            .iter()
+            .map(|pem| UnverifiedCertificate::try_from(*pem))
+            .collect::<Vec<_>>()
+            .into_iter()
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok(Self { certificates })
+    }
+}
+
+fn verify_certificate_not_revoked(
+    cert: &VerifiedCertificate,
+    trust_anchor: &VerifiedCertificate,
+    crls: &[UnverifiedCrl],
+    unix_time: Duration,
+) -> Result<()> {
+    for crl in crls {
+        if crl.issuer() == trust_anchor.subject_name() {
+            let verified_crl = crl.verify(&trust_anchor.public_key(), unix_time)?;
+            if verified_crl.is_cert_revoked(cert.serial_number()) {
+                return Err(Error::CertificateRevoked);
```

Comment:
Added a bullet item to #50, we're going to need more robust test data the current CRL has no revoked cert also need to better list out what we're planning on supporting from https://www.rfc-editor.org/rfc/rfc5280